pkg/secrets/secrets.go

```diff
@@ -0,0 +1,211 @@
+	if err != nil {
+		return nil, err
+	}
+	manager.lockfile = lock
+	manager.driver = driver
+	manager.secretsDBPath = filepath.Join(dirPath, secretsFile)
+	manager.db = new(db)
+	manager.db.Secrets = make(map[string]Secret)
+	manager.db.NameToID = make(map[string]string)
+	manager.db.IDToName = make(map[string]string)
+	return manager, nil
+}
+
+// Store creates a secret and stores it, given a name.
+// It stores secret metadata as well as the secret payload
+func (s *SecretsManager) Store(name string, data []byte) (string, error) {
+	err := validateSecretName(name)
+	if err != nil {
+		return "", err
+	}
+
+	s.lockfile.Lock()
+	defer s.lockfile.Unlock()
+
+	exist, err := s.secretExists(name)
+	if err != nil {
+		return "", err
+	}
+	if exist {
+		return "", errors.Wrapf(ErrSecretNameInUse, name)
+	}
+
+	secr := new(Secret)
+	secr.Name = name
+
+	for {
+		newID := stringid.GenerateNonCryptoID()
+		_, err := s.lookupSecret(newID)
+		if err != nil {
+			if errors.Cause(err) == ErrNoSuchSecret {
+				secr.ID = newID
+				break
+			} else {
+				return "", err
+			}
+		}
+	}
+
+	secr.Driver = s.driver.DriverType()
+	secr.Metadata = make(map[string]string)
+	secr.CreatedAt = time.Now()
+
+	err = s.driver.Store(secr.ID, data)
+	if err != nil {
+		return "", errors.Wrapf(err, "error creating secret %s", name)
+	}
+
+	err = s.store(secr)
+	if err != nil {
+		return "", errors.Wrapf(err, "error creating secret %s", name)
+	}
+
+	return secr.ID, nil
+}
+
+// Delete removes a secret
+// It removes secret metadata as well as the secret data
+func (s *SecretsManager) Delete(nameOrID string) (string, error) {
+	err := validateSecretName(nameOrID)
+	if err != nil {
+		return "", err
+	}
+
+	s.lockfile.Lock()
+	defer s.lockfile.Unlock()
+
+	_, id, err := s.getNameAndID(nameOrID)
+	if err != nil {
+		return "", err
+	}
+
+	err = s.driver.Delete(id)
+	if err != nil {
+		return "", errors.Wrapf(err, "error deleting secret %s", nameOrID)
+	}
+
+	err = s.delete(id)
+	if err != nil {
+		return "", errors.Wrapf(err, "error deleting secret %s", nameOrID)
+	}
+	return id, nil
+}
+
+// Lookup gives a secret's metadata
+func (s *SecretsManager) Lookup(namesOrIDs []string) ([]*Secret, error) {
+	if len(namesOrIDs) == 0 {
+		return nil, ErrInvalidSecretName
+	}
+	var lookups []*Secret
+	for _, nameOrID := range namesOrIDs {
+		secret, err := s.lookupSecret(nameOrID)
+		if err != nil {
+			return nil, err
+		}
+		lookups = append(lookups, secret)
+	}
+
+	return lookups, nil
+}
+
+// List lists all secrets
+func (s *SecretsManager) List() ([]Secret, error) {
+	s.lockfile.Lock()
+	defer s.lockfile.Unlock()
+
+	secrets, err := s.lookupAll()
+	if err != nil {
+		return nil, err
+	}
+	var ls []Secret
+	for _, v := range secrets {
+		ls = append(ls, v)
+
+	}
+	return ls, nil
+
+}
+
+// validateSecretName checks if the secret name is valid
+func validateSecretName(name string) error {
+	if name == "" {
+		return errors.Wrap(ErrInvalidSecretName, name)
```

Comment:
I think we should accept empty names. IDs _and_ names must be unique, and if we don't accept empty string users of this API had to find a unique name.

`secret create $data` and `secret create --name $name $data` should both work.

Reply:
If no name is specified, we'd auto-assign one.

Reply:
The Docker command line works like this `docker secret create secretname file` where both are required. There's no --name flag. Should we still allow empty names?

Reply:
No let's follow the Docker CLI exactly.

Reply:
Sounds good !